codec-redis/src/main/java/io/netty/handler/codec/redis/RedisDecoder.java

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.netty.handler.codec.redis;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.util.ByteProcessor;
+import io.netty.util.CharsetUtil;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.List;
+
+public class RedisDecoder extends ByteToMessageDecoder {
+
+    private static final int CRLF_LENGTH = 2;
+    private static final int NULL_RESPONSE = -1;
+
+    private Deque<DecodeState> stages;
+
+    private enum State {
+        DECODE_TYPE,
+        DECODE_INLINE, // SIMPLE_STRING, ERROR, INTEGER
+        DECODE_LENGTH, // BULK_STRING, ARRAY
+        DECODE_BULK_STRING,
+        DECODE_ARRAY,
+        DECODE_COMPLETE,
+    }
+
+    public RedisDecoder() {
+        resetDecoder();
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        try {
+            boolean next = true;
+            while (next) {
+                DecodeState current = stages.peek();
+                switch (current.state) {
+                case DECODE_TYPE:
+                    next = decodeType(current, in);
+                    break;
+                case DECODE_INLINE:
+                    next = decodeInline(current, in);
+                    break;
+                case DECODE_LENGTH:
+                    next = decodeLength(current, in);
+                    break;
+                case DECODE_BULK_STRING:
+                    next = decodeBulkStrings(current, in);
+                    break;
+                case DECODE_ARRAY:
+                    decodeArrays(current);
+                    break;
+                case DECODE_COMPLETE:
+                    aggregate(current, out);
```

Comment:
Redis responses can get really huge. While strings (well, byte chunks) are limited to 512MB, arrays are unbounded. Aggregation at this level leads to a `RedisMessage` that contains the huge array and requires the appropriate amount of memory.

How about splitting the protocol decoding and the message aggregation into two `Decoder`s? Users can decide whether they want to use the convenient way with aggregation or to follow more a streaming style of decoding.